core/include/gnuradio-4.0/Buffer.hpp

```diff
@@ -45,12 +45,36 @@ static_assert(std::is_same_v<double, value_type_t<std::array<double, 42>>>);
 
 } // namespace util
 
+// TODO: find a better place for these 3 concepts
+template<typename From, typename To>
+concept convertible_from_lvalue_to = std::convertible_to<const std::remove_cvref_t<From> &, To>;
+
+template<typename From, typename To>
+concept convertible_from_rvalue_to = std::convertible_to<std::remove_cvref_t<From> &&, To>;
+
+template<typename From, typename To>
+concept convertible_from_lvalue_only = convertible_from_lvalue_to<From, To> && !convertible_from_rvalue_to<From, To>;
+
+template<typename T>
+concept SpanLike = std::convertible_to<T, std::span<std::remove_cvref_t<typename T::value_type>>>;
+
+template<typename T>
+concept ConstSpanLike = std::convertible_to<T, std::span<const std::remove_cvref_t<typename T::value_type>>>;
+
+template<typename T>
+concept ConstSpanLvalueLike = convertible_from_lvalue_only<T, std::span<const std::remove_cvref_t<typename T::value_type>>>;
+
+template<typename T>
+concept ConsumableSpan = std::ranges::contiguous_range<T> && ConstSpanLvalueLike<T> && requires(T &s) { s.consume(0); };
+
+template<typename T>
+concept PublishableSpan = std::ranges::contiguous_range<T> && std::ranges::output_range<T, std::remove_cvref_t<typename T::value_type>> && SpanLike<T> && requires(T &s) { s.publish(0UZ); };
+
 // clang-format off
 // disable formatting until clang-format (v16) supporting concepts
 template<class T>
 concept BufferReader = requires(T /*const*/ t, const std::size_t n_items) {
-    { t.get(n_items) }     -> std::same_as<std::span<const util::value_type_t<T>>>;
-    { t.consume(n_items) } -> std::same_as<bool>;
+    { t.get(n_items) } ; // TODO: get returns CircularBuffer::buffer_reader::ConsumableInputRange
```

Comment:
N.B. could be replaced with the `ConsumableSpan` concept as the required return value.